### ISSUE TITLE
ArrayFire just doesn't work at all, Julia v0.5, v0.6

### ISSUE THREAD

**Issue by JMurph2015** [JuliaComputing/ArrayFire.jl#83]
Hi,
So I've rebuilt the ArrayFire library a couple times, compiled Julia v0.5 and v0.6 with and without CXX support, and none of it gets me anywhere beyond ArrayFire.jl just quitting out of the gate.  I can provide a lot more details, but to start, here's the output of several of the diagnostic commands.

```
murphyj@murphyj-Ubuntu:~$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.1026 (2016-10-18 23:57 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit bed6663 (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> using ArrayFire

julia> AFInfo()
ERROR: MethodError: no method matching throwAFerror()
Closest candidates are:
  throwAFerror(::Integer) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:48
 in af_info() at /home/murphyj/.julia/v0.6/ArrayFire/src/wrap.jl:1193
 in AFInfo() at /home/murphyj/.julia/v0.6/ArrayFire/src/utils.jl:93

julia> setBackend(AF_BACKEND_CPU)
true

julia> AFInfo()
ArrayFire v3.4.1 (CPU, 64-bit Linux, build b9055b1)
[0] Intel: Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz, 7848 MB, Max threads(8) 

julia> setBackend(AF_BACKEND_OPENCL)
true

julia> AFInfo()
ERROR: MethodError: no method matching throwAFerror()
Closest candidates are:
  throwAFerror(::Integer) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:48
 in af_info() at /home/murphyj/.julia/v0.6/ArrayFire/src/wrap.jl:1193
 in AFInfo() at /home/murphyj/.julia/v0.6/ArrayFire/src/utils.jl:93

julia> setBackend(AF_BACKEND_CUDA)
WARNING: bytestring(s::Cstring) is deprecated, use unsafe_string(s) instead.
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in bytestring(::Cstring) at ./deprecated.jl:50
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in af_set_backend(::UInt32) at /home/murphyj/.julia/v0.6/ArrayFire/src/wrap.jl:941
 in setBackend(::UInt32) at /home/murphyj/.julia/v0.6/ArrayFire/src/backend.jl:68
 in eval(::Module, ::Any) at ./boot.jl:236
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 in macro expansion at ./REPL.jl:97 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:65
while loading no file, in expression starting on line 0
ERROR: ArrayFire Error (501) : Failed to load dynamic library. 
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in af_set_backend(::UInt32) at /home/murphyj/.julia/v0.6/ArrayFire/src/wrap.jl:941
 in setBackend(::UInt32) at /home/murphyj/.julia/v0.6/ArrayFire/src/backend.jl:68
```

My system is on Ubuntu 16.04, with NVidia 960M graphics, NVidia CUDA toolkit 8.0 installed (includes nvidia-367), as well as nvidia-libopencl-367 installed.  Arrayfire is installed at /usr/local like usual.

```
murphyj@murphyj-Ubuntu:~$ echo $LD_LIBRARY_PATH 
/usr/local/include:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64
```

Any help would be much appreciated.
Thanks!

**Comment by pavanky** [comment]
Can you try building and running the arrayfire examples ?

**Comment by JMurph2015** [comment]
```
murphyj@murphyj-Ubuntu:~$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.1026 (2016-10-18 23:57 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit bed6663 (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> Pkg.test("ArrayFire")
INFO: Testing ArrayFire
WARNING: bytestring(s::Cstring) is deprecated, use unsafe_string(s) instead.
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in bytestring(::Cstring) at ./deprecated.jl:50
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in convert(::Type{ArrayFire.AFArray{Float32,2}}, ::Array{Float32,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:27
 in ArrayFire.AFArray{T,N}(::Array{Float32,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:31
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in process_options(::Base.JLOptions) at ./client.jl:271
 in _start() at ./client.jl:335
while loading /home/murphyj/.julia/v0.6/ArrayFire/test/runtests.jl, in expression starting on line 6
ERROR: LoadError: ArrayFire Error (998) : Internal error
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in convert(::Type{ArrayFire.AFArray{Float32,2}}, ::Array{Float32,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:27
 in ArrayFire.AFArray{T,N}(::Array{Float32,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:31
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in process_options(::Base.JLOptions) at ./client.jl:271
 in _start() at ./client.jl:335
while loading /home/murphyj/.julia/v0.6/ArrayFire/test/runtests.jl, in expression starting on line 6
=========================================================[ ERROR: ArrayFire ]=========================================================

failed process: Process(`/home/murphyj/julia/usr/bin/julia -Cnative -J/home/murphyj/julia/usr/lib/julia/sys.so --compile=yes --depwarn=yes --check-bounds=yes --code-coverage=none --color=yes --compilecache=yes /home/murphyj/.julia/v0.6/ArrayFire/test/runtests.jl`, ProcessExited(1)) [1]

======================================================================================================================================
ERROR: ArrayFire had test errors
 in #test#61(::Bool, ::Function, ::Array{AbstractString,1}) at ./pkg/entry.jl:749
 in (::Base.Pkg.Entry.#kw##test)(::Array{Any,1}, ::Base.Pkg.Entry.#test, ::Array{AbstractString,1}) at ./<missing>:0
 in (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#test,Tuple{Array{AbstractString,1}}})() at ./pkg/dir.jl:31
 in cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#test,Tuple{Array{AbstractString,1}}}, ::String) at ./file.jl:69
 in #cd#1(::Array{Any,1}, ::Function, ::Function, ::Array{AbstractString,1}, ::Vararg{Array{AbstractString,1},N}) at ./pkg/dir.jl:31
 in (::Base.Pkg.Dir.#kw##cd)(::Array{Any,1}, ::Base.Pkg.Dir.#cd, ::Function, ::Array{AbstractString,1}, ::Vararg{Array{AbstractString,1},N}) at ./<missing>:0
 in #test#3(::Bool, ::Function, ::String, ::Vararg{String,N}) at ./pkg/pkg.jl:258
 in test(::String, ::Vararg{String,N}) at ./pkg/pkg.jl:258
```

**Comment by JMurph2015** [comment]
Also I have this too.
```
murphyj@murphyj-Ubuntu:~/.julia/v0.6/ArrayFire/benchmarks$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.1026 (2016-10-18 23:57 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit bed6663 (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> using ArrayFire

julia> include("benchmark.jl")
WARNING: bytestring(s::Cstring) is deprecated, use unsafe_string(s) instead.
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in bytestring(::Cstring) at ./deprecated.jl:50
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in convert(::Type{ArrayFire.AFArray{Float64,2}}, ::Array{Float64,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:27
 in warmup() at /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl:4
 in benchmark() at /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl:58
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in eval(::Module, ::Any) at ./boot.jl:236
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 in macro expansion at ./REPL.jl:97 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:65
while loading /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl, in expression starting on line 107
ERROR: LoadError: ArrayFire Error (998) : Internal error
 in throwAFerror(::Int32) at /home/murphyj/.julia/v0.6/ArrayFire/src/error.jl:50
 in convert(::Type{ArrayFire.AFArray{Float64,2}}, ::Array{Float64,2}) at /home/murphyj/.julia/v0.6/ArrayFire/src/create.jl:27
 in warmup() at /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl:4
 in benchmark() at /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl:58
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
while loading /home/murphyj/.julia/v0.6/ArrayFire/benchmarks/benchmark.jl, in expression starting on line 107
```
I'll keep including the Julia header, just so that everyone can see the exact test conditions as best as possible.

**Comment by pavanky** [comment]
Sorry, I meant actual C++ examples.

**Comment by JMurph2015** [comment]
This seems to be promising, but /usr/local/cuda/lib64 has those files, and they are in the LD_LIBRARY_PATH
```
murphyj@murphyj-Ubuntu:~/src/examples/build$ make -j8
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/softmax_regression_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/naive_bayes_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'image_processing/pyramids_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/logistic_regression_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/knn_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/neural_network_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'machine_learning/rbm_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[  1%] Building CXX object CMakeFiles/example_softmax_regression_cuda.dir/machine_learning/softmax_regression.cpp.o
make[2]: *** No rule to make target '/usr/local/cuda/lib64/libcublas.so', needed by 'lin_algebra/qr_cuda'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[  1%] Building CXX object CMakeFiles/example_knn_cuda.dir/machine_learning/knn.cpp.o
[  1%] Building CXX object CMakeFiles/example_naive_bayes_cuda.dir/machine_learning/naive_bayes.cpp.o
[  1%] Building CXX object CMakeFiles/example_neural_network_cuda.dir/machine_learning/neural_network.cpp.o
[  1%] Building CXX object CMakeFiles/example_logistic_regression_cuda.dir/machine_learning/logistic_regression.cpp.o
[  1%] Building CXX object CMakeFiles/example_pyramids_cuda.dir/image_processing/pyramids.cpp.o
[  2%] Building CXX object CMakeFiles/example_qr_cuda.dir/lin_algebra/qr.cpp.o
[  2%] Building CXX object CMakeFiles/example_rbm_cuda.dir/machine_learning/rbm.cpp.o
CMakeFiles/Makefile2:252: recipe for target 'CMakeFiles/example_qr_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_qr_cuda.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:326: recipe for target 'CMakeFiles/example_pyramids_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_pyramids_cuda.dir/all] Error 2
CMakeFiles/Makefile2:215: recipe for target 'CMakeFiles/example_knn_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_knn_cuda.dir/all] Error 2
CMakeFiles/Makefile2:141: recipe for target 'CMakeFiles/example_naive_bayes_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_naive_bayes_cuda.dir/all] Error 2
CMakeFiles/Makefile2:289: recipe for target 'CMakeFiles/example_rbm_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_rbm_cuda.dir/all] Error 2
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/example_softmax_regression_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_softmax_regression_cuda.dir/all] Error 2
CMakeFiles/Makefile2:178: recipe for target 'CMakeFiles/example_logistic_regression_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_logistic_regression_cuda.dir/all] Error 2
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/example_neural_network_cuda.dir/all' failed
make[1]: *** [CMakeFiles/example_neural_network_cuda.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
murphyj@murphyj-Ubuntu:~/src/examples/build$ echo $LD_LIBRARY_PATH
/usr/local/cuda/lib64:/usr/local/cuda/lib:/usr/local/include:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64:/usr/local/lib:/usr/local/cuda-8.0/nvvm/lib64/:/usr/local/cuda-8.0/targets/x86_64-linux/lib/:/usr/local/cuda-8.0/lib64
```
Not sure what's going on, but if you have ideas, that'd be great (although I can move this on over to ArrayFire's Github or StackOverflow.)

**Comment by pavanky** [comment]
I think these may be two unrelated issues. But I'd suggest moving the build failure to arrayfire github.

**Comment by ranjanan** [comment]
@JMurph2015 Were you able to build ArrayFire? If you've moved the build issue to the arrayfire repo, could post a link to the tracker here?

**Comment by ranjanan** [comment]
@JMurph2015 can you please update this issue?

**Comment by ranjanan** [comment]
Closing this.

**Comment by JMurph2015** [comment]
Hi again,
I just saw the action on this issue.  Yes, I did find a standard way for ensuring that it ran properly each time.  I have a custom ldconfig file that I make sure to add to every machine I run it on.  It just makes sure that each part of Arrayfire + every part of the Cuda library and includes are linked.  Will post the exact file here when I'm at a workstation with it installed.
IIRC for whatever reason, the autolinking that should be recursive on things like /use/lib will only see things in that exact directory, but I don't think this is your guy's issue.
